client/host_queue.go

```diff
@@ -0,0 +1,265 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package client
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3db/interfaces/m3db"
+	"github.com/m3db/m3db/network/server/tchannelthrift/thrift/gen-go/rpc"
+
+	"github.com/uber/tchannel-go/thrift"
+)
+
+var (
+	errQueueClosed      = errors.New("host operation queue already closed")
+	errUnknownOperation = errors.New("unknown operation")
+)
+
+type hostQueue interface {
+	// Enqueue an operation
+	Enqueue(o op) error
+
+	// GetConnectionCount gets the current open connection count
+	GetConnectionCount() int
+
+	// Close the host queue, will flush any operations still pending
+	Close()
+}
+
+type queue struct {
+	sync.RWMutex
+
+	opts                  m3db.ClientOptions
+	host                  m3db.Host
+	connPool              connectionPool
+	writeBatchRequestPool writeBatchRequestPool
+	writeRequestArrayPool writeRequestArrayPool
+	size                  int
+	ops                   []op
+	opsArrayPool          opArrayPool
+	drainIn               chan []op
+	closed                bool
+}
+
+func newHostQueue(
+	host m3db.Host,
+	writeBatchRequestPool writeBatchRequestPool,
+	writeRequestArrayPool writeRequestArrayPool,
+	opts m3db.ClientOptions,
+) hostQueue {
+	size := opts.GetHostQueueOpsFlushSize()
```

Comment:
You mean for the `opArrayPool` capacity?  Yeah can make that a max of both of those.